src/types/messages.rs

```diff
@@ -0,0 +1,242 @@
+    /// The URL to the image.
+    pub url: String,
+    /// Metadata about the image referred to in url.
+    pub info: Option<ImageInfo>,
+    /// The URL to the thumbnail of the image.
+    pub thumbnail_url: Option<String>,
+    /// Metadata about the image referred to in thumbnail_url.
+    pub thumbnail_info: Option<ImageInfo>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Offer {
+    /// The type of session description. Must be 'offer'.
+    session_type: String,
+    /// The SDP text of the session description.
+    sdp: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Candidate {
+    /// Required. The SDP media type this candidate is intended for.
+    sdp_m_id: String,
+    /// Required. The index of the SDP 'm' line this candidate is intended for.
+    sdp_m_line_index: i32,
+    /// Required. The SDP 'a' line of the candidate.
+    candidate: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Answer {
+    // Required. The type of session description. Must be 'answer'.
+    session_type: String,
+    // Required. The SDP text of the session description.
+    sdp: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(untagged)]
+/// A message sent to a room.
+pub enum Message {
+    #[serde(rename="m.text")]
+    /// This message is the most basic message and is used to represent text.
+    Text{
+        /// The body of the message.
+        body: String,
+        /// must be m.text
+        msgtype: String,
+        /// The formatted body of the message (if the message is formatted).
+        #[serde(default)]
+        formatted_body: Option<String>,
+        /// The format of the formatted body (if the message is formatted).
+        #[serde(default)]
+        format: Option<String>
+    },
+    #[serde(rename="m.notice")]
+    /// A m.notice message should be considered similar to a plain m.text message except
+    /// that clients should visually distinguish it in some way.
+    /// It is intended to be used by automated clients, such as bots, bridges, and other
+    /// entities, rather than humans. Additionally, such automated agents which watch a
+    /// room for messages and respond to them ought to ignore m.notice messages. This
+    /// helps to prevent infinite-loop situations where two automated clients continuously
+    /// exchange messages, as each responds to the other.
+    Notice{
+        /// The notice text to send.
+        body: String,
+        /// must be m.notice
+        msgtype: String,
+        /// The formatted body of the message (if the message is formatted).
+        #[serde(default)]
+        formatted_body: Option<String>,
+        /// The format of the formatted body (if the message is formatted).
+        #[serde(default)]
+        format: Option<String>
+    },
+    #[serde(rename="m.image")]
+    /// This message represents a single image and an optional thumbnail.
+    /// Information about an image and it's thumbnail.
+    Image{
```

Comment:
You can be more DRY here by changing this to `Image(Image)`, since you made a new `Image` struct with these attributes...